### PR TITLE
vendor shading filter from vispy

### LINKED ISSUE #2817: New surface shaders are broken
## 🐛 Bug

With `0.4.9rc0`, new shaders for mesh rendering were added (#2792).
On my machine, this broke surface layers:

## To Reproduce
```python
import napari
import numpy as np
pos = np.array([[0, 0, 0], [1, 1, 1], [0, 1, 0]]) * 10
faces = np.array([[0, 1, 2]])
v = napari.Viewer(ndisplay=3)
v.add_surface((pos, faces))
```
This fails with a [long trace](https://pastebin.com/ZjCNNRfi), whose starting point seems to be:
```python
File "~/.local/lib/python3.9/site-packages/napari/_vispy/vendored/filters/mesh.py", line 301, in _update_data
    if not self._attached:
AttributeError: 'ShadingFilter' object has no attribute '_attached'
```
and ends with
```python
RuntimeError: Cannot draw program if code has not been set
```

## Environment

```yaml
napari: 0.4.9rc0
Platform: Linux-5.12.8-arch1-1-x86_64-with-glibc2.33
System: Arch Linux
Python: 3.9.5 (default, May 24 2021, 12:50:35) [GCC 11.1.0]
Qt: 5.15.2
PyQt5: 5.15.4
NumPy: 1.20.3
SciPy: 1.6.3
Dask: 2020.12.0
VisPy: 0.6.6

OpenGL:
- GL version: 4.6.0 NVIDIA 465.31
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1080, scale 1.0
- screen 2: resolution 1920x1200, scale 1.0

Plugins:
- animation: 0.0.2
- console: 0.0.3
- mrcfile-reader: 0.1.2
- napari-em-reader: 0.1.0
- napari-properties-plotter: 0.1.dev2+g2734460.d20210520
- ndtiffs: 0.1.1
- scikit-image: 0.4.8
- svg: 0.1.5
```

### PR DESCRIPTION
**Pull request by alisterburt**
Partial fix for #2817 by vendoring the current version of the shading filter class `Filter` from vispy
together with #2820 this should fix #2817 completely

@brisvag if you could test the combination of these two PRs on your system that would be great, I can't reproduce your error locally

### CODE REVIEW

**Comment by alisterburt** [comment]
just pushed a temporary solution before I head off for the evening, not happy with it and would rather hide the shading controls in 2D completely but it's at least not completely broken 🙂

**Comment by goanpeca** [comment]
Still a small conflict :)

**Comment by alisterburt** [comment]
sorry @goanpeca - should be okay now

**Comment by goanpeca** [comment]
Hmm it seems that rebase went weird 🙃

**Comment by alisterburt** [comment]
**curses git**

**Comment by alisterburt** [comment]
back in working state, thanks for your patience!

File: napari/layers/surface/surface.py [hunk 1]
```diff
@@ -309,6 +312,10 @@ def shading(self):
 
     @shading.setter
     def shading(self, shading):
+        if self._ndisplay < 3:
+            raise ValueError(
```
Comment (sofroniewn) [inline napari/layers/surface/surface.py]: Can we maybe make this a warning and then just overwrite `self._shading = 'none'`. I feel like i'd rather just warn and fix it then raise an error

**Review by sofroniewn — APPROVED** [review]
LGTM! I think its reasonable for shading to just have no effect in 2D, i'm fine with always overwriting to `'none'` and leaving combo box for now

This has fixed all outstanding issues for me. Maybe @brisvag can confirm all works for him too and then we can merge!!

**Comment by brisvag** [comment]
This fixed my problem! I think we can merge and close #2817 !

**Comment by sofroniewn** [comment]
I have created https://github.com/alisterburt/napari/pull/2 that would fix the tests here and fix an additional 2D to 3D to 2D bug. cc @alisterburt @brisvag

**Comment by alisterburt** [comment]
thanks guys

@sofroniewn I didn't want to merge your PR with the weird diff so added those changes myself rather than do battle with git

On my machine there's actually still a (non-fatal) bug if you run `examples/add_surface_2D` and toggle from 2D -> 3D -> 2D, I'm guessing you don't see this?

<details>
<summary>traceback</summary>
<br>

```python
ValueError                                Traceback (most recent call last)
~/Programming/napari/napari/utils/action_manager.py in <lambda>()
     51     def callable(self, context):
     52         if not hasattr(self, '_command_with_context'):
---> 53             self._command_with_context = lambda: call_with_context(
        global self._command_with_context = undefined
        global call_with_context = <function call_with_context at 0x1503e80d0>
        global self.command = undefined
        global context = undefined
     54                 self.command, context
     55             )
~/Programming/napari/napari/utils/action_manager.py in call_with_context(function=<function toggle_ndisplay>, context=<napari.utils.action_manager.Context object>)
     40     ]
     41     ctx = {k: v for k, v in context.items() if k in context_keys}
---> 42     return function(**ctx)
        function = <function toggle_ndisplay at 0x150e659d0>
        ctx = {'viewer': Viewer(axes=Axes(visible=False, labels=True, colored=True, dashed=False, arrows=True), camera=Camera(center=(0.0, 5.0, 10.0), zoom=47.3575, angles=(0.0, 0.0, 90.0), perspective=0.0, interactive=True), cursor=Cursor(position=(10.294694624008583, -22.13942085242857), scaled=True, size=1, style='standard'), dims=Dims(ndim=2, ndisplay=3, last_used=0, range=((0.0, 10.0, 1.0), (0.0, 20.0, 1.0)), current_step=(0, 0), order=(0, 1), axis_labels=('0', '1')), grid=GridCanvas(enabled=False, stride=1, shape=(-1, -1)), layers=[<Surface layer 'Surface' at 0x1682617f0>], scale_bar=ScaleBar(visible=False, colored=False, ticks=True, position='bottom_right', font_size=10.0, unit=None), text_overlay=TextOverlay(visible=False, color=array([1., 1., 1., 1.]), font_size=10.0, position='top_left', text=''), help='', status='Surface [ 10 -22]', theme='dark', title='napari', mouse_move_callbacks=[], mouse_drag_callbacks=[], mouse_wheel_callbacks=[<function dims_scroll at 0x150e3edc0>], _persisted_mouse_event={}, _mouse_drag_gen={}, _mouse_wheel_gen={}, keymap={'Control-Shift-C': <bound method QtViewer.toggle_console_visibility of <napari._qt.qt_viewer.QtViewer object at 0x150e82a60>>})}
     43 
     44 
~/Programming/napari/napari/components/_viewer_key_bindings.py in toggle_ndisplay(viewer=Viewer(axes=Axes(visible=False, labels=True, col...._qt.qt_viewer.QtViewer object at 0x150e82a60>>}))
     46         viewer.dims.ndisplay = 2
     47     else:
---> 48         viewer.dims.ndisplay = 3
        viewer.dims.ndisplay = 3
     49 
     50 
~/Programming/napari/napari/utils/events/evented_model.py in __setattr__(self=Dims(ndim=2, ndisplay=3, last_used=0, range=((0....tep=(0, 0), order=(0, 1), axis_labels=('0', '1')), name='ndisplay', value=3)
    153         are_equal = self.__eq_operators__.get(name, operator.eq)
    154         if not are_equal(after, before):
--> 155             getattr(self.events, name)(value=after)  # emit event
        global getattr = undefined
        self.events = <napari.utils.events.event.EmitterGroup object at 0x150e7a610>
        name = 'ndisplay'
        value = 3
        after = 3
    156 
    157     # expose the private EmitterGroup publically
~/Programming/napari/napari/utils/events/event.py in __call__(self=<napari.utils.events.event.EventEmitter object>, *args=(), **kwargs={'value': 3})
    571                     continue
    572 
--> 573                 self._invoke_callback(cb, event)
        self._invoke_callback = <bound method EventEmitter._invoke_callback of <napari.utils.events.event.EventEmitter object at 0x150e7a670>>
        cb = <bound method ViewerModel._update_layers of Viewer(axes=Axes(visible=False, labels=True, colored=True, dashed=False, arrows=True), camera=Camera(center=(0.0, 5.0, 10.0), zoom=47.3575, angles=(0.0, 0.0, 90.0), perspective=0.0, interactive=True), cursor=Cursor(position=(10.294694624008583, -22.13942085242857), scaled=True, size=1, style='standard'), dims=Dims(ndim=2, ndisplay=3, last_used=0, range=((0.0, 10.0, 1.0), (0.0, 20.0, 1.0)), current_step=(0, 0), order=(0, 1), axis_labels=('0', '1')), grid=GridCanvas(enabled=False, stride=1, shape=(-1, -1)), layers=[<Surface layer 'Surface' at 0x1682617f0>], scale_bar=ScaleBar(visible=False, colored=False, ticks=True, position='bottom_right', font_size=10.0, unit=None), text_overlay=TextOverlay(visible=False, color=array([1., 1., 1., 1.]), font_size=10.0, position='top_left', text=''), help='', status='Surface [ 10 -22]', theme='dark', title='napari', mouse_move_callbacks=[], mouse_drag_callbacks=[], mouse_wheel_callbacks=[<function dims_scroll at 0x150e3edc0>], _persisted_mouse_event={}, _mouse_drag_gen={}, _mouse_wheel_gen={}, keymap={'Control-Shift-C': <bound method QtViewer.toggle_console_visibility of <napari._qt.qt_viewer.QtViewer object at 0x150e82a60>>})>
        event = <Event blocked=False handled=False native=None source=None sources=[] type=ndisplay>
    574                 if event.blocked:
    575                     break
~/Programming/napari/napari/utils/events/event.py in _invoke_callback(self=<napari.utils.events.event.EventEmitter object>, cb=<bound method ViewerModel._update_layers of View..._qt.qt_viewer.QtViewer object at 0x150e82a60>>})>, event=<Event blocked=False handled=False native=None source=None sources=[] type=ndisplay>)
    593             cb(event)
    594         except Exception:
--> 595             _handle_exception(
        global _handle_exception = <function _handle_exception at 0x14d1918b0>
        self.ignore_callback_errors = False
        self.print_callback_errors = 'reminders'
        self = <napari.utils.events.event.EventEmitter object at 0x150e7a670>
        global cb_event = undefined
        cb = <bound method ViewerModel._update_layers of Viewer(axes=Axes(visible=False, labels=True, colored=True, dashed=False, arrows=True), camera=Camera(center=(0.0, 5.0, 10.0), zoom=47.3575, angles=(0.0, 0.0, 90.0), perspective=0.0, interactive=True), cursor=Cursor(position=(10.294694624008583, -22.13942085242857), scaled=True, size=1, style='standard'), dims=Dims(ndim=2, ndisplay=3, last_used=0, range=((0.0, 10.0, 1.0), (0.0, 20.0, 1.0)), current_step=(0, 0), order=(0, 1), axis_labels=('0', '1')), grid=GridCanvas(enabled=False, stride=1, shape=(-1, -1)), layers=[<Surface layer 'Surface' at 0x1682617f0>], scale_bar=ScaleBar(visible=False, colored=False, ticks=True, position='bottom_right', font_size=10.0, unit=None), text_overlay=TextOverlay(visible=False, color=array([1., 1., 1., 1.]), font_size=10.0, position='top_left', text=''), help='', status='Surface [ 10 -22]', theme='dark', title='napari', mouse_move_callbacks=[], mouse_drag_callbacks=[], mouse_wheel_callbacks=[<function dims_scroll at 0x150e3edc0>], _persisted_mouse_event={}, _mouse_drag_gen={}, _mouse_wheel_gen={}, keymap={'Control-Shift-C': <bound method QtViewer.toggle_console_visibility of <napari._qt.qt_viewer.QtViewer object at 0x150e82a60>>})>
        event = <Event blocked=False handled=False native=None source=None sources=[] type=ndisplay>
    596                 self.ignore_callback_errors,
    597                 self.print_callback_errors,
~/Programming/napari/napari/utils/events/event.py in _invoke_callback(self=<napari.utils.events.event.EventEmitter object>, cb=<bound method ViewerModel._update_layers of View..._qt.qt_viewer.QtViewer object at 0x150e82a60>>})>, event=<Event blocked=False handled=False native=None source=None sources=[] type=ndisplay>)
    591     def _invoke_callback(self, cb: Callback, event: Event):
    592         try:
--> 593             cb(event)
        cb = <bound method ViewerModel._update_layers of Viewer(axes=Axes(visible=False, labels=True, colored=True, dashed=False, arrows=True), camera=Camera(center=(0.0, 5.0, 10.0), zoom=47.3575, angles=(0.0, 0.0, 90.0), perspective=0.0, interactive=True), cursor=Cursor(position=(10.294694624008583, -22.13942085242857), scaled=True, size=1, style='standard'), dims=Dims(ndim=2, ndisplay=3, last_used=0, range=((0.0, 10.0, 1.0), (0.0, 20.0, 1.0)), current_step=(0, 0), order=(0, 1), axis_labels=('0', '1')), grid=GridCanvas(enabled=False, stride=1, shape=(-1, -1)), layers=[<Surface layer 'Surface' at 0x1682617f0>], scale_bar=ScaleBar(visible=False, colored=False, ticks=True, position='bottom_right', font_size=10.0, unit=None), text_overlay=TextOverlay(visible=False, color=array([1., 1., 1., 1.]), font_size=10.0, position='top_left', text=''), help='', status='Surface [ 10 -22]', theme='dark', title='napari', mouse_move_callbacks=[], mouse_drag_callbacks=[], mouse_wheel_callbacks=[<function dims_scroll at 0x150e3edc0>], _persisted_mouse_event={}, _mouse_drag_gen={}, _mouse_wheel_gen={}, keymap={'Control-Shift-C': <bound method QtViewer.toggle_console_visibility of <napari._qt.qt_viewer.QtViewer object at 0x150e82a60>>})>
        event = <Event blocked=False handled=False native=None source=None sources=[] type=ndisplay>
    594         except Exception:
    595             _handle_exception(
~/Programming/napari/napari/components/viewer_model.py in _update_layers(self=Viewer(axes=Axes(visible=False, labels=True, col...._qt.qt_viewer.QtViewer object at 0x150e82a60>>}), event=<Event blocked=False handled=False native=None source=None sources=[] type=ndisplay>, layers=[<Surface layer 'Surface' at 0x1682617f0>])
    279         layers = layers or self.layers
    280         for layer in layers:
--> 281             layer._slice_dims(
        layer._slice_dims = <bound method Layer._slice_dims of <Surface layer 'Surface' at 0x1682617f0>>
        self.dims.point = (0.0, 0.0)
        self.dims.ndisplay = 3
        self.dims.order = (0, 1)
    282                 self.dims.point, self.dims.ndisplay, self.dims.order
    283             )
~/Programming/napari/napari/layers/base/base.py in _slice_dims(self=<Surface layer 'Surface'>, point=[0.0, 0.0], ndisplay=3, order=[0, 1])
    899         # Update the point values
    900         self._dims_point = point[offset:]
--> 901         self._update_dims()
        self._update_dims = <bound method Layer._update_dims of <Surface layer 'Surface' at 0x1682617f0>>
    902         self._set_editable()
    903 
~/Programming/napari/napari/layers/base/base.py in _update_dims(self=<Surface layer 'Surface'>, event=None)
    568         self._ndim = ndim
    569 
--> 570         self.refresh()
        self.refresh = <bound method Layer.refresh of <Surface layer 'Surface' at 0x1682617f0>>
    571 
    572     @property
~/Programming/napari/napari/layers/base/base.py in refresh(self=<Surface layer 'Surface'>, event=None)
    971         if self.visible:
    972             self.set_view_slice()
--> 973             self.events.set_data()
        self.events.set_data = <napari.utils.events.event.EventEmitter object at 0x168261190>
    974             self._update_thumbnail()
    975             self._set_highlight(force=True)
~/Programming/napari/napari/utils/events/event.py in __call__(self=<napari.utils.events.event.EventEmitter object>, *args=(), **kwargs={})
    571                     continue
    572 
--> 573                 self._invoke_callback(cb, event)
        self._invoke_callback = <bound method EventEmitter._invoke_callback of <napari.utils.events.event.EventEmitter object at 0x168261190>>
        cb = <bound method VispySurfaceLayer._on_data_change of <napari._vispy.vispy_surface_layer.VispySurfaceLayer object at 0x168248790>>
        event = <Event blocked=False handled=False native=None source=None sources=[] type=set_data>
    574                 if event.blocked:
    575                     break
~/Programming/napari/napari/utils/events/event.py in _invoke_callback(self=<napari.utils.events.event.EventEmitter object>, cb=<bound method VispySurfaceLayer._on_data_change ...py.vispy_surface_layer.VispySurfaceLayer object>>, event=<Event blocked=False handled=False native=None source=None sources=[] type=set_data>)
    593             cb(event)
    594         except Exception:
--> 595             _handle_exception(
        global _handle_exception = <function _handle_exception at 0x14d1918b0>
        self.ignore_callback_errors = False
        self.print_callback_errors = 'reminders'
        self = <napari.utils.events.event.EventEmitter object at 0x168261190>
        global cb_event = undefined
        cb = <bound method VispySurfaceLayer._on_data_change of <napari._vispy.vispy_surface_layer.VispySurfaceLayer object at 0x168248790>>
        event = <Event blocked=False handled=False native=None source=None sources=[] type=set_data>
    596                 self.ignore_callback_errors,
    597                 self.print_callback_errors,
~/Programming/napari/napari/utils/events/event.py in _invoke_callback(self=<napari.utils.events.event.EventEmitter object>, cb=<bound method VispySurfaceLayer._on_data_change ...py.vispy_surface_layer.VispySurfaceLayer object>>, event=<Event blocked=False handled=False native=None source=None sources=[] type=set_data>)
    591     def _invoke_callback(self, cb: Callback, event: Event):
    592         try:
--> 593             cb(event)
        cb = <bound method VispySurfaceLayer._on_data_change of <napari._vispy.vispy_surface_layer.VispySurfaceLayer object at 0x168248790>>
        event = <Event blocked=False handled=False native=None source=None sources=[] type=set_data>
    594         except Exception:
    595             _handle_exception(
~/Programming/napari/napari/_vispy/vispy_surface_layer.py in _on_data_change(self=<napari._vispy.vispy_surface_layer.VispySurfaceLayer object>, event=<Event blocked=False handled=False native=None source=None sources=[] type=set_data>)
     53             vertices = np.pad(vertices, ((0, 0), (0, 1)))
     54 
---> 55         self._on_shading_change()
        self._on_shading_change = <bound method VispySurfaceLayer._on_shading_change of <napari._vispy.vispy_surface_layer.VispySurfaceLayer object at 0x168248790>>
     56         self.node.set_data(
     57             vertices=vertices, faces=faces, vertex_values=vertex_values
~/Programming/napari/napari/_vispy/vispy_surface_layer.py in _on_shading_change(self=<napari._vispy.vispy_surface_layer.VispySurfaceLayer object>, event=None)
     96             self.node.shading = None
     97         else:
---> 98             self.node.shading = self.layer.shading
        self.node.shading = 'flat'
        self.layer.shading = 'flat'
     99         self.node.mesh_data_changed()
    100 
~/Programming/vispy/vispy/util/frozen.py in __setattr__(self=<Mesh>, key='shading', value='flat')
     15                                  '"unfreeze()" to allow addition of new '
     16                                  'attributes' % (key, self))
---> 17         object.__setattr__(self, key, value)
        global object.__setattr__ = undefined
        self = <Mesh at 0x1682692b0>
        key = 'shading'
        value = 'flat'
     18 
     19     def freeze(self):
~/Programming/napari/napari/_vispy/vendored/mesh.py in shading(self=<Mesh>, shading='flat')
    160             self.shading_filter = ShadingFilter(shading=shading)
    161         else:
--> 162             self.shading_filter.shading = shading
        self.shading_filter.shading = 'flat'
        shading = 'flat'
    163         self.attach(self.shading_filter)
    164 
~/Programming/napari/napari/_vispy/vendored/filters/mesh.py in shading(self=<napari._vispy.vendored.filters.mesh.ShadingFilter object>, shading='flat')
    223         assert shading in (None, 'flat', 'smooth')
    224         self._shading = shading
--> 225         self._update_data()
        self._update_data = <bound method ShadingFilter._update_data of <napari._vispy.vendored.filters.mesh.ShadingFilter object at 0x168269ee0>>
    226 
    227     @property
~/Programming/napari/napari/_vispy/vendored/filters/mesh.py in _update_data(self=<napari._vispy.vendored.filters.mesh.ShadingFilter object>)
    310         self.fshader['shading_enabled'] = 1 if self._shading is not None else 0
    311         normals = self._visual.mesh_data.get_vertex_normals(indexed='faces')
--> 312         self._normals.set_data(normals, convert=True)
        self._normals.set_data = <bound method DataBuffer.set_data of <VertexBuffer size=6 last_dim=3>>
        normals = array([[[ 1.,  1.],
        [ 1.,  1.],
        [ 1.,  1.]],
       [[ 1.,  1.],
        [ 1.,  1.],
        [-1., -1.]]], dtype=float32)
        global convert = undefined
    313 
    314     def on_mesh_data_updated(self, event):
~/Programming/vispy/vispy/gloo/buffer.py in set_data(self=<VertexBuffer size=6 last_dim=3>, data=array([[[ 1.,  1.],
        [ 1.,  1.],
        ... [ 1.,  1.],
        [-1., -1.]]], dtype=float32), copy=False, **kwargs={'convert': True})
    184             Additional arguments.
    185         """
--> 186         data = self._prepare_data(data, **kwargs)
        data = array([[[ 1.,  1.],
        [ 1.,  1.],
        [ 1.,  1.]],
       [[ 1.,  1.],
        [ 1.,  1.],
        [-1., -1.]]], dtype=float32)
        self._prepare_data = <bound method VertexBuffer._prepare_data of <VertexBuffer size=6 last_dim=3>>
        kwargs = {'convert': True}
    187         self._dtype = data.dtype
    188         # This works around some strange NumPy bug where a float32 array
~/Programming/vispy/vispy/gloo/buffer.py in _prepare_data(self=<VertexBuffer size=6 last_dim=3>, data=array([[[ 1.,  1.],
        [ 1.,  1.],
        ... [ 1.,  1.],
        [-1., -1.]]], dtype=float32), convert=True)
    433                 c = 1
    434             if self._last_dim and c != self._last_dim:
--> 435                 raise ValueError('Last dimension should be %s not %s'
        global ValueError = undefined
        self._last_dim = 3
        c = 2
    436                                  % (self._last_dim, c))
    437             dtype_def = ('f0', data.dtype.base)
ValueError: Last dimension should be 3 not 2```

</details>

**Comment by sofroniewn** [comment]
> On my machine there's actually still a (non-fatal) bug if you run examples/add_surface_2D and toggle from 2D -> 3D -> 2D, I'm guessing you don't see this?

I saw that before my fix, I'll look again though, it was def a little finicky

File: napari/_vispy/vispy_surface_layer.py [hunk 2]
```diff
@@ -81,6 +85,17 @@ def _on_gamma_change(self, event=None):
     def _on_shading_change(self, event=None):
         if self.layer.shading == 'none':
             self.node.shading = None
+            if self.node.shading_filter is not None:
+                self.node.shading_filter._attached = False
```
Comment (sofroniewn) [inline napari/_vispy/vispy_surface_layer.py:89]: @alisterburt I wonder if we should contribute these two lines upstream into vispy so they get called when `self.node.shading = None`. Would be nice to have have to go into their private stuff

**Review by sofroniewn — COMMENTED** [review]
@alisterburt I think you missed one more `_attached = False` from my PR. I just added it back. Seems to have fixed the 2D -> 3D -> 2D issue for me locally, let's see if all tests now pass too. See my comment above about maybe contributing that change upstream too

**Comment by sofroniewn** [comment]
All tests passing now, I think this is good to go!!

File: napari/_vispy/vispy_surface_layer.py (hunk 2 above)
Comment (alisterburt) [reply napari/_vispy/vispy_surface_layer.py:89]: great idea, will do!

**Pull request merged**